docs/element-definitions.md

```diff
@@ -1353,6 +1356,339 @@ the `inferred` link tells the user that the element was created based on some
+                  }
+                ]
+              }
+            }
+          ]
+        }
+      ]
+    }
+  ]
+}
+```
+
+### Token Authentication Scheme (Base API Element)
+
+This describes an authentication scheme that uses a token as a way to authenticate and authorize. The token MAY exist as an HTTP header field or a query parameter.
+
+- One of
+    - `httpHeaderName` (string)
+    - `queryParameterName` (string)
+
+When used as a query parameter, an HREF Variable is not required to be defined within the scope of the resource or transition, but is rather infered from the used token authentications scheme.
+
+#### Properties
+
+- `element`: Token Authentication Scheme (string, fixed)
+- `content` (array[Member Element])
+
+#### Example
+
+This example shows a custom token authentication scheme being defined as `Custom Token Auth` that uses a query parameter for the token. This scheme is then used on an HTTP transaction within a resource. Please note this example is incomplete for the sake of keeping it short.
+
+```json
+{
+  "element": "category",
+  "meta": {
+    "classes": ["api"]
+  },
+  "content": [
+    {
+      "element": "category",
+      "meta": {
+        "classes": ["authSchemes"]
+      },
+      "content": [
+        {
+          "element": "Token Authentication Scheme",
+          "meta": {
+            "id": "Custom Token Auth"
+          },
+          "content": [
+            {
+              "element": "member",
+              "content": {
+                "key": {
+                  "element": "string",
+                  "content": "queryParameterName"
+                },
+                "value": {
+                  "element": "string",
+                  "content": "api_key"
+                }
+              }
+            }
+          ]
+        }
+      ]
+    },
+    {
+      "element": "resource",
+      "attributes": {
+        "href": "/users"
+      },
+      "content": [
+        {
+          "element": "transition",
+          "content": [
+            {
+              "element": "httpTransaction",
+              "attributes": {
+                "authSchemes": [
+                  {
+                    "element": "Custom Token Auth"
+                  }
+                ]
+              }
+            }
+          ]
+        }
+      ]
+    }
+  ]
+}
+```
+
+### Oauth2 Scheme (Base API Element)
+
+This describes an authentication scheme that uses Oauth2 as defined in [RFC 6749](https://tools.ietf.org/html/rfc6749).
+
+The element MAY have the following members to define additional information about the Oauth2 implementation.
+
+- `scopes` (array[string])
+- `grantType` (enum)
+    - `authorization code`
+    - `implicit`
+    - `resource owner password credentials`
+    - `client credentials`
+
+Transition elements are used to define the URLs for the authorize and token endpoints for the Oauth2 schemes. When including these endpoints, the following link relations SHOULD be used.
```

Comment:
I would love to include a way for describing the parameters/attributes and their semantic meaning so that tools can utilise this information when exposing OAuth authentication flow.

For example, if a tool would want to allow a user to login to this website. It would need to use the `authorize` relation, but the tool doesn't understand the information it will need to send when performing this transition.

For example, the client doesn't know how to send the redirect uri in the OAuth authorise flow, the redirect URI key name could differ based on implementations. For example, one API might use `redirect_uri`, another might use `RedirectURI`, `redirectUri`, etc.

This is not only important for tools, but also humans reading an API documentation described in Refract. For example, if I describe my API authentication flow based on OAuth in an API Blueprint, I would expect to be able to describe the detains on how it works along with showing this in the documentation to users trying to implement clients against my API.

I've tried to solve this problem inside the [API Blueprint RFC 5](https://github.com/apiaryio/api-blueprint-rfcs/blob/master/rfcs/0005-authentication-oauth2.md#endpoints) where a user can describe the parameters or attributes using a data structure. Where the type of attribute is used to identify the semantic meaning behind the attribute.

For example, `Redirect Callback` type in a data structure would allow you to declare how the redirect callback (URI) is sent to the API.

Adding a data structure would allow the user to properly document how the transition works along with the parameters a client may send. This can also include whether certain parameters are required, for example `state` or `scope`.

```apib
#### Authorization [GET /authorize]

+ Relation: authorize
+ Request
    + Parameters
        + `response_type`: code (Response Type, required)
        + `client_id`: b58051708627fa6a5b7e49108f3c7574 (Client ID, required)
        + state (State)
        + scope (Scope)
        + `redirect_uri` (Redirect Callback, required)

+ Redirect
    + Parameters
        + code (Code)
        + state (State)
```

It may also be important to include how the redirect callback works, what fields we are expected to get back and what they key names are.